develop/items/custom-data-components.md

```diff
@@ -0,0 +1,271 @@
+
+![Tooltip showing "Used 8 times"](/assets/develop/items/custom_component_3.png)
+
+## Removing component value {#removing-component-value}
+
+You can also remove a component from your `ItemStack` if it is no longer needed. This is done by using the `remove()` method which takes your component type.
+
+```java
+stack.remove(ModComponents.CLICK_COUNT_COMPONENT);
+```
+
+This method also returns the value of the component before being removed, so you can also use it like so:
+
+```java
+int oldCount = stack.remove(ModComponents.CLICK_COUNT_COMPONENT);
+```
+
+## Advanced Data Components {#advanced-data-components}
+
+You may need to store multiple attributes in a single component. As a vanilla example, the `minecraft:food` component stores several values related to food, such as `nutrition`, `saturation`, `eat_seconds` and more. In this guide we'll refer to them as "composite" components.
+
+For composite components, you must create a `record` class to store the data. This is the type we'll register in our component type and what we'll read and write when interacting with an `ItemStack`. Start by making a new record class in the `component` package we made earlier.
+
+```java
+public record MyCustomComponent() {
+    
+}
+```
+
+Notice that there's a set of brackets after the class name. This is where we define the list of properties we want our component to have. Let's add a float and a boolean called `temperature` and `burnt` respectively.
+
+@[code transcludeWith=::1](@/reference/latest/src/main/java/com/example/docs/component/MyCustomComponent.java)
+
+Since we are defining a custom data structure there won't be a pre-existing `Codec` for our use case like with the [basic component](#basic-data-components). This means we're going to have to construct our own codec. Let's define one in our record class using a `RecordCodecBuilder` which we can reference once we register the component. For more details on using a `RecordCodecBuilder` you can refer to [this section of the Codecs page](../codecs.md#merging-codecs-for-record-like-classes).
+
+@[code transcludeWith=::2](@/reference/latest/src/main/java/com/example/docs/component/MyCustomComponent.java)
+
+You can see that we are defining a list of custom fields based on the primitive `Codec` types. However we are also telling it what our fields are called using `fieldOf()`, and then using `forGetter()` to tell the game which attribute of our record to populate.
+
+You can also define optional fields by using `optionalFieldOf()` and passing a default value as the second argument. Any fields not marked optional will be required when setting the component using `/give` so make sure you mark any optional arguments as such when creating your codec.
+
+Finally we call `apply()` and pass our record's constructor. For more details on how to construct codecs and more advanced use cases be sure to read the [Codecs](../codecs.md) page.
```

Comment:
Seems like all of this is already explained in the Codecs page. Is it worth repeating it here?

Reply:
I think so, users are much more likely to know about and search for a "custom components" page than a "codec" page which new devs and regular players would have no idea about.

Reply:
Yup this is fine, if users want to do more advanced stuff they can look at the wider codec page.